domains/client/domain-operator/src/bundle_processor.rs

```diff
@@ -72,30 +93,154 @@ where
     }
 }
 
-struct ReceiptValidator<Client> {
+struct ReceiptValidator<Client, Block, CBlock, CClient, Backend, E, ParentChain, ParentChainBlock> {
+    domain_id: DomainId,
     client: Arc<Client>,
+    fraud_proof_generator: FraudProofGenerator<Block, CBlock, Client, CClient, Backend, E>,
+    parent_chain: ParentChain,
+    _phantom: std::marker::PhantomData<ParentChainBlock>,
 }
 
-impl<Client> Clone for ReceiptValidator<Client> {
+impl<Client, Block, CBlock, CClient, Backend, E, ParentChain, ParentChainBlock> Clone
+    for ReceiptValidator<Client, Block, CBlock, CClient, Backend, E, ParentChain, ParentChainBlock>
+where
+    ParentChain: Clone,
+    ParentChainBlock: BlockT,
+{
     fn clone(&self) -> Self {
         Self {
+            domain_id: self.domain_id,
             client: self.client.clone(),
+            fraud_proof_generator: self.fraud_proof_generator.clone(),
+            parent_chain: self.parent_chain.clone(),
+            _phantom: self._phantom,
         }
     }
 }
 
-impl<Client> ReceiptValidator<Client> {
-    pub fn new(client: Arc<Client>) -> Self {
-        Self { client }
+impl<Client, Block, CBlock, CClient, Backend, E, ParentChain, ParentChainBlock>
+    ReceiptValidator<Client, Block, CBlock, CClient, Backend, E, ParentChain, ParentChainBlock>
+{
+    pub fn new(
+        domain_id: DomainId,
+        client: Arc<Client>,
+        fraud_proof_generator: FraudProofGenerator<Block, CBlock, Client, CClient, Backend, E>,
+        parent_chain: ParentChain,
+    ) -> Self {
+        Self {
+            domain_id,
+            client,
+            fraud_proof_generator,
+            parent_chain,
+            _phantom: PhantomData,
+        }
     }
 }
 
-impl<Block, CBlock, Client> domain_block_preprocessor::ValidateReceipt<Block, CBlock>
-    for ReceiptValidator<Client>
+impl<Client, Block, CBlock, CClient, Backend, E, ParentChain, ParentChainBlock>
+    ReceiptValidator<Client, Block, CBlock, CClient, Backend, E, ParentChain, ParentChainBlock>
 where
     Block: BlockT,
     CBlock: BlockT,
-    Client: AuxStore,
+    Client:
+        HeaderBackend<Block> + BlockBackend<Block> + AuxStore + ProvideRuntimeApi<Block> + 'static,
+    Client::Api: sp_block_builder::BlockBuilder<Block>
+        + sp_api::ApiExt<Block, StateBackend = StateBackendFor<Backend, Block>>,
+    CClient: HeaderBackend<CBlock> + 'static,
+    Backend: sc_client_api::Backend<Block> + Send + Sync + 'static,
+    TransactionFor<Backend, Block>: sp_trie::HashDBT<HashFor<Block>, sp_trie::DBValue>,
+    E: CodeExecutor,
+    ParentChainBlock: BlockT,
+    ParentChain: ParentChainInterface<Block, ParentChainBlock>,
+{
+    fn check_invalid_bundles_field(
```

Comment:
This is much more complicated than I expected and I'm not fully reading/understanding the whole logic TBH.

What we need to do here is simply to **find the first mismatch item of the `invalid_bundles` list** between the local receipt and the external receipt. Differences in either `bundle_index` or `invalid_bundle_type` should be considered as mismatch.

The following is my proposal I think it is more simply and straight forward but let me know if I have missed anything:

```rust
pub fn check_receipt_invalid_bundles(
    &self,
    local_receipt: ExecutionReceipt,
    external_receipt: ExecutionReceipt,
) -> Option<FraudProof> {
    if local_receipt.invalid_bundles == external_receipt.invalid_bundles {
        return None;
    }
    for (local_invalid_bundle, external_invalid_bundle) in local_receipt
        .invalid_bundles
        .zip(external_receipt.invalid_bundles.iter())
    {
        if local_invalid_bundle != external_invalid_bundle {
            if local_invalid_bundle.invalid_bundle_type
                != external_invalid_bundle.external_invalid_bundle
            {
                // TODO: proving `external_invalid_bundle` is an invalid bundle of type `local_invalid_bundle.invalid_bundle_type`
            }
            // FIXME: wee need to add a check to the consensus chain runtime to ensure for all the ER included in the consensus block
            // the `bundle_index` field of `ER.invalid_bundles` must be strictly increasing
            match local_invalid_bundle
                .bundle_index
                .cmp(external_invalid_bundle.bundle_index)
            {
                Ordering::Greater => {
                    // TODO: proving `external_invalid_bundle` is not an invalid of type `external_invalid_bundle.invalid_bundle_type`
                }
                Ordering::Less => {
                    // TODO: proving `local_invalid_bundle` is indeed an invalid bundle of type `local_invalid_bundle.invalid_bundle_type`
                }
                Ordering::Equal => unreachable!("checked above; qed"),
            }
        }
    }
    match local_receipt
        .invalid_bundles
        .len()
        .cmp(external_receipt.invalid_bundles.len())
    {
        Ordering::Greater => {
            let invalid_bundle =
                local_receipt.invalid_bundles[external_receipt.invalid_bundles.len()];
            // TODO: proving `invalid_bundle` is indeed an invalid bundle of type `invalid_bundle.invalid_bundle_type`
        }
        Ordering::Less => {
            let valid_bundle =
                external_receipt.invalid_bundles[local_receipt.invalid_bundles.len()];
            // TODO: proving `valid_bundle` is not an invalid bundle of type `valid_bundle.invalid_bundle_type`
        }
        Ordering::Equal => unreachable!("checked above; qed"),
    }
}
```

Reply:
This is complicated because it creates fraud proof for all the discrepancy in the `invalid_bundles` type.

Reply:
But, Let me still try to simplify it.